packages/components/src/resize-tooltip/utils.js

```diff
@@ -0,0 +1,229 @@
+	onMove = noop,
+	onResize = noop,
+	position = POSITIONS.cursor,
+	showPx = false,
+} ) {
+	const [ isDragging, setIsDragging ] = useState( false );
+	const [ isActive, setIsActive ] = useState( false );
+	/**
+	 * The width/height values derive from this special useResizeAwere hook.
+	 * This custom hook uses injects an iFrame into the element, allowing it
+	 * to tap into the onResize (window) callback events.
+	 */
+	const [ resizeListener, sizes ] = useResizeAware();
+
+	/**
+	 * The moveX and moveY values are used to track whether the label should
+	 * display width, height, or width x height.
+	 */
+	const [ moveX, setMoveX ] = useState( false );
+	const [ moveY, setMoveY ] = useState( false );
+
+	/**
+	 * Cached dimension values to check for width/height updates from the
+	 * sizes property from useResizeAware()
+	 */
+	const heightRef = useRef( height );
+	const widthRef = useRef( width );
+
+	/**
+	 * This timeout is used with setMoveX and setMoveY to determine of
+	 * both width and height values have changed at (roughly) the same time.
+	 */
+	const moveTimeoutRef = useRef();
+
+	const { width, height } = sizes;
+
+	useEffect( () => {
+		/**
+		 * On the initial render of useResizeAware, the height and width values are
+		 * null. They are calculated then set using via an internal useEffect hook.
+		 */
+		const isRendered = width !== null || height !== null;
+
+		if ( ! isRendered ) return;
+
+		const didWidthChange = width !== widthRef.current;
+		const didHeightChange = height !== heightRef.current;
+
+		if ( ! didWidthChange && ! didHeightChange ) return;
+
+		if ( didWidthChange ) {
+			setMoveX( true );
+		}
+
+		if ( didHeightChange ) {
+			setMoveY( true );
+		}
+
+		onResize( { width, height } );
+	}, [ width, height ] );
+
+	useEffect( () => {
+		const handleOnMouseDown = () => {
+			if ( moveTimeoutRef.current ) {
+				clearTimeout( moveTimeoutRef.current );
+			}
+			setIsDragging( true );
+			setMoveX( false );
+			setMoveY( false );
+		};
+
+		const handleOnMouseUp = () => {
+			setIsDragging( false );
+			setIsActive( false );
+
+			if ( moveTimeoutRef.current ) {
+				clearTimeout( moveTimeoutRef.current );
+			}
+
+			moveTimeoutRef.current = setTimeout( () => {
+				setMoveX( false );
+				setMoveY( false );
+			}, fadeTimeout * 2 );
+		};
+
+		const handleOnMouseMove = ( event ) => {
+			if ( ! isDragging ) return;
+
+			if ( width !== widthRef.current || height !== heightRef.current ) {
+				widthRef.current = width;
+				heightRef.current = height;
+
+				onMove( event );
+				setIsActive( true );
+			}
+		};
+
+		document.addEventListener( 'mousedown', handleOnMouseDown );
+		document.addEventListener( 'mousemove', handleOnMouseMove );
+		document.addEventListener( 'mouseup', handleOnMouseUp );
+
+		return () => {
+			document.removeEventListener( 'mousedown', handleOnMouseDown );
+			document.removeEventListener( 'mousemove', handleOnMouseMove );
+			document.removeEventListener( 'mouseup', handleOnMouseUp );
```

Comment:
Just looking at the code, I don't think this cleanup ever does anything. That's because the `handle*` functions are re-instantiated every time the effect runs.

The other thing is that I believe we could be more judicious about the dependencies of this effect. By refactoring the handler functions and properly looking at what values are needed, I believe we could reduce the `[ width, height, isActive, isDragging ]` dependency list, and therefore not tear down and reset this effect so often.

Reply:
Thank you for pointing this out! I just pushed up some changes which should feel better (from a code/useEffect perspective). Let me know what you think of these revisions 🙏

Reply:
> Just looking at the code, I don't think this cleanup ever does anything. That's because the `handle*` functions are re-instantiated every time the effect runs.

I don't why I said this, as it didn't make sense: the registered listeners are properly referenced in the tear-down function returned by the effect, so we should always be able to remove the old event listeners.

Reply:
> Let me know what you think of these revisions

So I've been going back and forth looking at this effect and feeling generally unsatisfied when pondering things like the dependencies, and I think I know why. We're dealing with effects of first degree and effects of second degree, so to speak: the first-degree effects are the registration of the `mouse*` event listeners; the second-degree ones are the listeners themselves. This makes the reasoning about dependencies more confusing.

If we were stricter about constructing these effects, we might refactor this as follows:

```js
export function useResizeLabel( … ) {
  // ...

  const clearMoveTimeout = useCallback( () => {
    if ( moveTimeoutRef.current ) {
      clearTimeout( moveTimeoutRef.current );
    }
  }, [ moveTimeoutRef ] );

  const handleOnMouseDown = useCallback( () => {
    if ( ! isCursor ) return;

    setIsDragging( true );
    clearMoveTimeout();
    unsetMoveXY();
  }, [
    // meaningful dependencies (mutable)
    isCursor,
    // formal dependencies (assumed constant)
    setIsDragging, clearMoveTimeout, unsetMoveYX,
  ] );

  const handleOnMouseMove = useCallback( /* same as above */ );

  const handleOnMouseUp = useCallback( /* same as above */ );
  
  useEffect( () => {
    document.addEventListener( 'mousedown', handleOnMouseDown );
    return () => {
      document.removeEventListener( 'mousedown', handleOnMouseDown );
    };
  }, [ handleOnMouseDown ] );

  useEffect( /* same as above for `mousemove` */ );

  useEffect( /* same as above for `mouseup` */ );

  return { ... };
}
```

Note those comments about "meaningful" vs. "formal" dependencies. That was my of not making a decision as to whether we should be a strict as possible when listing dependencies or not. :) I'll leave that up to you.

Reply:
Thanks for the detailed feedback @mcsf !

I'm still unsure of when `useEffect`/`useCallback` dependencies should be used. I use to be in the camp of over-using `useCallback` (and would have done what you had suggested).

However, after reading Kent C Dodd's posts on `useCallback`, I've eased up a lot:
https://kentcdodds.com/blog/usememo-and-usecallback

Nowadays, I'm more in the mindset of...
Use `useCallback` if it causes an obvious problem (too many executions, infinite loops, slow perf). Otherwise, it should be okay.

That being said! I'm happy to make changes to accommodate the preferred styles of the project 🙇

Reply:
> I'm happy to make changes to accommodate the preferred styles of the project

First, a disclaimer: this is really not a preferred style, just the result of my spitballing, and it's important to remember that in general, when it comes to React hooks, much is up in the air! I don't think you should just change to what I suggest. :)

I totally agree on being careful about `useCallback`, and in that aspect my example is really not compelling.

My focus was on isolating each effect as much as possible and then being _exhaustive_ about its dependencies. I use that term based on the idea of the [`exhaustive-deps` lint rule](https://github.com/facebook/react/issues/14920). The idea is quite interesting in theory.

In my mind, it can prevent a lot of subtle bugs that could happen, especially in the long term as components are tweaked and it becomes harder to trace the implicit (i.e. not exhaustively declared) dependencies of an effect or other hook.

In practice, though, it becomes awkward. Any function called inside an effect or callback is technically a dependency, and if we're dealing with a function declared in the body of the component then this dependency changes all the time. At that point, we have three choices:

1. Drop the requirement for "exhaustive" deps
2. Generously use `useCallback` to cache those functions so that the effect in question doesn't re-run for no reason
3. Accept that the effect will run frequently

This obviously only opens the door for more questions. :) Like, for choices 2 and 3, how much is too much function caching, and how much it too much cache busting? For choice 1, how can design hooks so as to minimise bugs related to state, cache and closure?

* * *

Finally, and to help unblock things here: I really don't mind that we write effects in the way that makes most intuitive sense and is performant. In the case of these effects, it's clear that the the point is to register three DOM-event handlers that will more or less be active for as long as the component exists. We can totally keep the effects as you wrote them. ;)